cranelift/2023/cranelift-08-09.md

```diff
@@ -11,12 +11,58 @@
 1. Announcements
     1. _Submit a PR to add your announcement here_
 1. Other agenda items
-    1. MachBuffer and branch optimizations (acrichto)
+    1. MachBuffer and branch optimizations (alexcrichton)
     1. _Submit a PR to add your item here_
 
 ## Notes
 
+1. MachBuffer and branch optimization (alexcrichton) https://github.com/bytecodealliance/wasmtime/issues/6798
+- Issue on slow compilation.
+    - Not full module, but alexcrichton was able to reproduce from screenshot. A label fixup was quadratic by accident.  First thought: binary heap, worked for that aspect, but did not work for optimizations to the MachBuffer. Original issue fixed by cfallin’s PR changes handling of islands. Might want to in the future stop eagerly promoting branches. 
+    - Two problems: linearizing the CFG, and branch/block optimizations. Extended basic blocks are single-in, multiple out. Using these gives some new optimization opportunity, but leads to complexity/changes for standard compiler technology. ~4 years ago Cranelift used this, now moved away from that, Cranelift has an actual CFG including VCode.
+        - Question: can we do branch optimizations in the mid end? E.g. branch constant folding. Nothing precluding us from converting an if to a jump. 
+        - alexcrichton question: the MachBuffer does nice linearization, does it do more than that? Answer: yes, it does 3 optimizations: jump to fall through to just jump, jump over conditional converted to inverted conditional, redirect branch to unconditional branch to remove indirection (plus cleanup if unused). Design deciscion in regalloc2, require all critical edges to be split. 
```

Comment:
"jump over conditional converted to inverted conditional" -> "conditional jump over unconditional jump converted to inverted conditional"